### PR TITLE
Http error callback (Catch http error statuscode)

### PR DESCRIPTION
**Pull request by manuelblum**
Catching and handling of HTTP Errors (Client/Server Error Codes)

We often have the requirement in our apps to handle certain http errors/status codes. 

For example, if a specific url returns a "Statuscode 401 - Not Authorised". This is the indicator that the users access token has been expired or has been revoked. So if we receive a 401 status code, we try to refresh the access token via api call or logout the user on refresh failure.

Another use case could be to present a native error view if a server error (Statuscode: 500) happens.

Eventually `onError` and `onHttpError` could be combined to `onError` but we explicitly decided that we do not want to do this, because:
- Handling http errors and network errors should be two different things.
- Native implementation (ios and android) has also separated catching status code into separate functions.
    - `-(void)webView:(WKWebView *)webView decidePolicyForNavigationResponse:(WKNavigationResponse *)navigationResponse decisionHandler:(void (^)(WKNavigationResponsePolicy))decisionHandler` (iOS)
    - `void onReceivedHttpError(WebView webView, WebResourceRequest request, WebResourceResponse errorResponse)` (Android)

We think that this feature is not that uncommon and we are looking forward to see this feature being shipped. What do you think about it?

Feature works on iOS and Android >= 21 (Lolipop). As far as i see, there is no way to support Android < 21.

Closes Issue #189

### CODE REVIEW

File: ios/RNCWKWebView.m [hunk 1]
```diff
@@ -718,6 +719,25 @@ - (void)                  webView:(WKWebView *)webView
   decisionHandler(WKNavigationResponsePolicyAllow);
 }
 
+-(void)webView:(WKWebView *)webView decidePolicyForNavigationResponse:(WKNavigationResponse *)navigationResponse decisionHandler:(void (^)(WKNavigationResponsePolicy))decisionHandler
```
Comment (TheAlmightyBob) [inline ios/RNCWKWebView.m]: `decidePolicyForNavigationResponse` will also be invoked for nested frames. Could this handler avoid reporting those or provide the info with the event?

I've been using `isTopFrame = [response.URL isEqual:webView.URL];` in a local patch to detect this, based on the idea that by the time this is invoked `webView.URL` should have already been updated to be the main document URL.

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java [hunk 2]
```diff
@@ -627,6 +630,24 @@ public void onReceivedError(
         new TopLoadingErrorEvent(webView.getId(), eventData));
     }
 
+    @Override
+    public void onReceivedHttpError(WebView webView, WebResourceRequest request, WebResourceResponse errorResponse) {
```
Comment (TheAlmightyBob) [inline android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: Could this use `request.isForMainFrame()` to filter out errors from nested frames or include with the JS event whether or not it's for the main frame?

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java [hunk 3]
```diff
@@ -627,6 +630,24 @@ public void onReceivedError(
...
+      super.onReceivedHttpError(webView, request, errorResponse);
+      int statusCode = 0;
+      if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
+        statusCode = errorResponse.getStatusCode();
+        if(statusCode >= 400 && statusCode < 600)
+        {
+          WritableMap eventData = createWebViewEvent(webView, request.getUrl().toString());
+          eventData.putDouble("statusCode", statusCode);
```
Comment (TheAlmightyBob) [inline android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: Might be helpful to include `errorResponse.getReasonPhrase()` as well?

**Comment by TheAlmightyBob** [comment]
+1 to keeping `onError` and `onHttpError` separate. For example, in some cases it may be desirable to just go ahead and show the HTTP error page from the server rather than hiding it.

**Comment by jamonholmgren** [comment]
Looks like there's a small conflict, @manuelblum, can you resolve?

**Comment by Titozzz** [comment]
Plus answering all the good point @TheAlmightyBob made 😀

**Comment by avertin** [comment]
This is something that would be super helpful for my project as well. +1 on this PR

**Comment by calmyournerves** [comment]
Would still be great to have this in, any chance to make it happen @manuelblum?

**Pull request closed without merging**